File: LeetCode/543. 二叉树的直径.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: TreeNode) -> int:
        diameter = 0
        def depth(node: TreeNode) -> int:
            if not node:
                return 0
            left = depth(node.left)
            right = depth(node.right)
            return max(left, right) + 1
        def traverse(node: TreeNode):
            nonlocal diameter
            if not node:
                return
            left = depth(node.left)
            right = depth(node.right)
            diameter = max(diameter, left + right)
            traverse(node.left)
            traverse(node.right)
        traverse(root)
        return diameter

    
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTreeInfer(self, root: TreeNode) -> int:
        diameter = 0
        def depth(node: TreeNode) ->int:
            nonlocal diameter
            if not node:
                return 0
            left = depth(node.left)
            right = depth(node.right)
            diameter = max(left + right, diameter)
            return max(left, right) + 1
        depth(root)
        return diameter

File: LeetCode/543. 二叉树的直径.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function diameterOfBinaryTree(root: TreeNode | null): number {
    let result = 0;
    const depthMap = new Map<TreeNode, number>();
    const depth = (node: TreeNode | null): number => {
        if (node === null) return 0;
        if (depthMap.has(node)) return depthMap.get(node);
        const left = depth(node.left);
        const right = depth(node.right);
        const d = Math.max(left, right) + 1;
        depthMap.set(node, d);
        return d;
    };
    const traverse = (node: TreeNode | null) => {
        if (node === null) return;
        const left = depth(node.left);
        const right = depth(node.right);
        const cur = left + right;
        result = Math.max(cur, result);
        traverse(node.left);
        traverse(node.right);
    };
    traverse(root);
    return result;
};


/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function diameterOfBinaryTreeInfer(root: TreeNode | null): number {
    let diameter = 0;
    const depth = (node: TreeNode | null): number => {
        if(node === null) return 0;
        const left = depth(node.left);
        const right = depth(node.right);
        diameter = Math.max(diameter, left + right);
        return Math.max(left, right) + 1;
    };
    depth(root);
    return diameter;
};
